hudhook/src/hooks/common.rs

```diff
@@ -0,0 +1,119 @@
+use imgui::*;
+use parking_lot::MutexGuard;
+use windows::Win32::Foundation::{HWND, LPARAM, LRESULT, WPARAM};
+use windows::Win32::UI::Input::KeyboardAndMouse::*;
+use windows::Win32::UI::WindowsAndMessaging::{WHEEL_DELTA, WM_XBUTTONDBLCLK, XBUTTON1, *};
+
+pub(crate) type WndProcType =
+    unsafe extern "system" fn(hwnd: HWND, umsg: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT;
+
+pub(crate) trait ImguiRendererCommon {
+    fn io_mut(&mut self) -> &mut imgui::Io;
+    fn set_focus(&mut self, focus: bool);
+    fn is_focus(&self) -> bool;
+    fn get_wnd_proc(&self) -> WndProcType;
+
+    fn init_io(&mut self) {
+        let mut io = ImguiRendererCommon::io_mut(self);
+
+        io.nav_active = true;
+        io.nav_visible = true;
+
+        // Initialize keys
+        io[Key::Tab] = VK_TAB.0 as _;
+        io[Key::LeftArrow] = VK_LEFT.0 as _;
+        io[Key::RightArrow] = VK_RIGHT.0 as _;
+        io[Key::UpArrow] = VK_UP.0 as _;
+        io[Key::DownArrow] = VK_DOWN.0 as _;
+        io[Key::PageUp] = VK_PRIOR.0 as _;
+        io[Key::PageDown] = VK_NEXT.0 as _;
+        io[Key::Home] = VK_HOME.0 as _;
+        io[Key::End] = VK_END.0 as _;
+        io[Key::Insert] = VK_INSERT.0 as _;
+        io[Key::Delete] = VK_DELETE.0 as _;
+        io[Key::Backspace] = VK_BACK.0 as _;
+        io[Key::Space] = VK_SPACE.0 as _;
+        io[Key::Enter] = VK_RETURN.0 as _;
+        io[Key::Escape] = VK_ESCAPE.0 as _;
+        io[Key::A] = VK_A.0 as _;
+        io[Key::C] = VK_C.0 as _;
+        io[Key::V] = VK_V.0 as _;
+        io[Key::X] = VK_X.0 as _;
+        io[Key::Y] = VK_Y.0 as _;
+        io[Key::Z] = VK_Z.0 as _;
+    }
+}
+
+#[must_use]
+pub(crate) fn imgui_wnd_proc_impl(
+    hwnd: HWND,
+    umsg: u32,
+    WPARAM(wparam): WPARAM,
+    LPARAM(lparam): LPARAM,
+    mut imgui_renderer: MutexGuard<Box<impl ImguiRendererCommon>>,
+) -> LRESULT {
+    let mut io = imgui_renderer.io_mut();
+    match umsg {
+        WM_KEYDOWN | WM_SYSKEYDOWN => {
+            if wparam < 256 {
+                io.keys_down[wparam as usize] = true;
+            }
+        },
+        WM_KEYUP | WM_SYSKEYUP => {
+            if wparam < 256 {
+                io.keys_down[wparam as usize] = false;
+            }
+        },
+        WM_LBUTTONDOWN | WM_LBUTTONDBLCLK => {
+            io.mouse_down[0] = true;
+        },
+        WM_RBUTTONDOWN | WM_RBUTTONDBLCLK => {
+            io.mouse_down[1] = true;
+        },
+        WM_MBUTTONDOWN | WM_MBUTTONDBLCLK => {
+            io.mouse_down[2] = true;
+        },
+        WM_XBUTTONDOWN | WM_XBUTTONDBLCLK => {
+            let btn = if super::hiword(wparam as _) == XBUTTON1.0 as u16 { 3 } else { 4 };
```

Comment:
As a general principle, I'd prefer having these functions imported rather than referring to them by the full path.